File: examples/kotlin/projector-receipt/src/main/kotlin/dev/angzarr/examples/projector/ReceiptProjectorLogic.kt
package dev.angzarr.examples.projector

import dev.angzarr.EventBook
import dev.angzarr.Projection
import com.google.protobuf.Any
import examples.Domains.*

/**
 * Interface for receipt projector business logic.
 */
interface ReceiptProjectorLogic {
    fun buildState(eventBook: EventBook): ReceiptState
    fun formatReceipt(state: ReceiptState, orderId: String): String
    fun createProjection(eventBook: EventBook, projectorName: String): Projection?
}

/**
 * Default implementation of receipt projector logic for Order domain.
 */
class DefaultReceiptProjectorLogic : ReceiptProjectorLogic {

    override fun buildState(eventBook: EventBook): ReceiptState {
        var state = ReceiptState.empty()

        for (page in eventBook.pagesList) {
            val event = page.event ?: continue
            state = applyEvent(state, event)
        }

        return state
    }

    private fun applyEvent(state: ReceiptState, event: Any): ReceiptState {
        return when {
            event.`is`(OrderCreated::class.java) -> {
                val e = event.unpack(OrderCreated::class.java)
                state.copy(
                    customerId = e.customerId,
                    items = e.itemsList,
                    subtotalCents = e.subtotalCents,
                    discountCents = e.discountCents
                )
            }
            event.`is`(LoyaltyDiscountApplied::class.java) -> {
                val e = event.unpack(LoyaltyDiscountApplied::class.java)
                state.copy(
                    loyaltyPointsUsed = e.pointsUsed,
                    discountCents = state.discountCents + e.discountCents
                )
            }
            event.`is`(PaymentSubmitted::class.java) -> {
                val e = event.unpack(PaymentSubmitted::class.java)
                state.copy(
                    paymentMethod = e.paymentMethod,
                    finalTotalCents = e.amountCents
                )
            }
            event.`is`(OrderCompleted::class.java) -> {
                val e = event.unpack(OrderCompleted::class.java)
                state.copy(
                    completedAt = e.completedAt,
                    isCompleted = true
                )
            }
            else -> state
        }
    }

    override fun formatReceipt(state: ReceiptState, orderId: String): String {
        val sb = StringBuilder()
        val line = "═".repeat(40)
        val thinLine = "─".repeat(40)

        val shortOrderId = if (orderId.length > 16) orderId.take(16) else orderId
        val shortCustId = if (state.customerId.length > 16) state.customerId.take(16) else state.customerId

        sb.appendLine(line)
        sb.appendLine("           RECEIPT")
        sb.appendLine(line)
        sb.appendLine("Order: $shortOrderId...")
        sb.appendLine("Customer: ${if (shortCustId.isEmpty()) "N/A" else "$shortCustId..."}")
        sb.appendLine(thinLine)

        for (item in state.items) {
            val lineTotal = item.quantity * item.unitPriceCents
            sb.appendLine("${item.quantity} x ${item.name} @ $${String.format("%.2f", item.unitPriceCents / 100.0)} = $${String.format("%.2f", lineTotal / 100.0)}")
        }

        sb.appendLine(thinLine)
        sb.appendLine("Subtotal:              $${String.format("%.2f", state.subtotalCents / 100.0)}")

        if (state.discountCents > 0) {
            val discountType = if (state.loyaltyPointsUsed > 0) "loyalty" else "coupon"
            sb.appendLine("Discount ($discountType):       -$${String.format("%.2f", state.discountCents / 100.0)}")
        }

        sb.appendLine(thinLine)
        sb.appendLine("TOTAL:                 $${String.format("%.2f", state.finalTotalCents / 100.0)}")
        sb.appendLine("Payment: ${state.paymentMethod}")
        sb.appendLine(thinLine)

        val loyaltyPointsEarned = (state.finalTotalCents / 100) * ReceiptState.POINTS_PER_DOLLAR
        sb.appendLine("Loyalty Points Earned: $loyaltyPointsEarned")
        sb.appendLine(line)
        sb.appendLine("     Thank you for your purchase!")
        sb.append(line)

        return sb.toString()
    }

    override fun createProjection(eventBook: EventBook, projectorName: String): Projection? {
        val state = buildState(eventBook)
        if (!state.isCompleted) return null

        val orderId = eventBook.cover?.root?.value?.toByteArray()
            ?.joinToString("") { "%02x".format(it) } ?: ""

        val loyaltyPointsEarned = (state.finalTotalCents / 100) * ReceiptState.POINTS_PER_DOLLAR

        val receipt = Receipt.newBuilder()
            .setOrderId(orderId)
            .setCustomerId(state.customerId)
            .addAllItems(state.items)
            .setSubtotalCents(state.subtotalCents)
            .setDiscountCents(state.discountCents)
            .setFinalTotalCents(state.finalTotalCents)
            .setPaymentMethod(state.paymentMethod)
            .setLoyaltyPointsEarned(loyaltyPointsEarned)
            .setFormattedText(formatReceipt(state, orderId))
            .build()

        return Projection.newBuilder()
            .setCover(eventBook.cover)
            .setProjector(projectorName)
            .setSequence(eventBook.pagesList.size)
            .setProjection(Any.pack(receipt))
            .build()
    }
}

File: examples/ruby/projector-receipt/lib/receipt_projector_logic.rb
# frozen_string_literal: true

require 'time'

# Pure business logic for receipt projector.
# No gRPC dependencies - can be tested in isolation.
module ReceiptProjectorLogic
  POINTS_PER_DOLLAR = 10

  ReceiptData = Struct.new(
    :customer_id,
    :items,
    :subtotal_cents,
    :discount_cents,
    :loyalty_points_used,
    :final_total_cents,
    :payment_method,
    :loyalty_points_earned,
    :completed_at,
    keyword_init: true
  )

  class << self
    def project(event_book)
      return nil if event_book.nil? || event_book.pages.empty?

      customer_id = ''
      items = []
      subtotal_cents = 0
      discount_cents = 0
      loyalty_points_used = 0
      final_total_cents = 0
      payment_method = ''
      completed_at = nil
      is_completed = false

      event_book.pages.each do |page|
        next unless page.event

        type_url = page.event.type_url

        if type_url.end_with?('OrderCreated')
          event = Examples::OrderCreated.decode(page.event.value)
          customer_id = event.customer_id
          items = event.items.to_a
          subtotal_cents = event.subtotal_cents
          discount_cents = event.discount_cents
        elsif type_url.end_with?('LoyaltyDiscountApplied')
          event = Examples::LoyaltyDiscountApplied.decode(page.event.value)
          loyalty_points_used = event.points_used
          discount_cents += event.discount_cents
        elsif type_url.end_with?('PaymentSubmitted')
          event = Examples::PaymentSubmitted.decode(page.event.value)
          final_total_cents = event.amount_cents
          payment_method = event.payment_method
        elsif type_url.end_with?('OrderCompleted')
          event = Examples::OrderCompleted.decode(page.event.value)
          completed_at = event.completed_at
          is_completed = true
        end
      end

      return nil unless is_completed

      loyalty_points_earned = (final_total_cents / 100) * POINTS_PER_DOLLAR

      ReceiptData.new(
        customer_id: customer_id,
        items: items,
        subtotal_cents: subtotal_cents,
        discount_cents: discount_cents,
        loyalty_points_used: loyalty_points_used,
        final_total_cents: final_total_cents,
        payment_method: payment_method,
        loyalty_points_earned: loyalty_points_earned,
        completed_at: completed_at
      )
    end

    def format_receipt(receipt_data, order_id: nil)
      timestamp = receipt_data.completed_at ? Time.at(receipt_data.completed_at.seconds).strftime('%Y-%m-%d %H:%M:%S') : Time.now.strftime('%Y-%m-%d %H:%M:%S')

      short_order_id = order_id && order_id.length > 16 ? order_id[0...16] : order_id
      short_customer_id = receipt_data.customer_id.length > 16 ? receipt_data.customer_id[0...16] : receipt_data.customer_id

      lines = []
      lines << '=' * 40
      lines << 'RECEIPT'.center(40)
      lines << '=' * 40
      lines << ''
      lines << "Order: #{short_order_id}..." if short_order_id
      lines << "Customer: #{short_customer_id.empty? ? 'N/A' : "#{short_customer_id}..."}"
      lines << "Date: #{timestamp}"
      lines << ''
      lines << '-' * 40

      receipt_data.items.each do |item|
        line_total = item.quantity * item.unit_price_cents
        lines << "#{item.quantity} x #{item.name} @ $#{format_cents(item.unit_price_cents)} = $#{format_cents(line_total)}"
      end

      lines << '-' * 40
      lines << "Subtotal: $#{format_cents(receipt_data.subtotal_cents)}".rjust(40)

      if receipt_data.discount_cents > 0
        discount_type = receipt_data.loyalty_points_used > 0 ? 'loyalty' : 'coupon'
        lines << "Discount (#{discount_type}): -$#{format_cents(receipt_data.discount_cents)}".rjust(40)
      end

      lines << '-' * 40
      lines << "TOTAL: $#{format_cents(receipt_data.final_total_cents)}".rjust(40)
      lines << "Payment: #{receipt_data.payment_method}"
      lines << '-' * 40
      lines << "Loyalty Points Earned: #{receipt_data.loyalty_points_earned}"
      lines << '=' * 40
      lines << 'Thank you for your purchase!'.center(40)
      lines << '=' * 40

      lines.join("\n")
    end

    private

    def format_cents(cents)
      format('%.2f', cents / 100.0)
    end
  end
end
